### PR TITLE
Code cleanup

### PR DESCRIPTION
**Pull request by andersy005**
This PR introduces small changes to enhance code maintainability...

- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Review by max-sixty — APPROVED** [review]
Nice! Thanks @andersy005

File: xarray/backends/scipy_.py
Comment (mathause) [inline xarray/backends/scipy_.py:211]:
```suggestion
                f"unexpected encoding for scipy backend: {list(variable.encoding)}" 
```

File: xarray/coding/strings.py
Comment (mathause) [inline xarray/coding/strings.py:176]:
```suggestion
            "cannot stack dask character array with "
```

File: xarray/core/accessor_dt.py
Comment (mathause) [inline xarray/core/accessor_dt.py:347]: Also this one I find more difficult to parse in my head than the old version.

File: xarray/core/alignment.py
Comment (mathause) [inline xarray/core/alignment.py:339]: Nice!

File: xarray/core/coordinates.py
Comment (mathause) [inline xarray/core/coordinates.py:388]:
```suggestion
                f"dimension coordinate {k!r} conflicts between "
                f"indexed and indexing objects:\n{obj[k]}\nvs.\n{coords[k]}"
```

File: xarray/core/utils.py
Comment (mathause) [inline xarray/core/utils.py:283]:
```suggestion
        raise ValueError(f"the first argument to {func_name} must be a dictionary")
```
(not sure what the `"."` is/ was doing there - i.e. if it's needed)

File: xarray/core/utils.py
Comment (mathause) [inline xarray/core/utils.py:286]:
```suggestion
            f"cannot specify both keyword and positional arguments to {func_name}"
```
dido concerning the `"."`

File: xarray/plot/utils.py
Comment (mathause) [inline xarray/plot/utils.py:53]: (I don't like the pattern so much but you could remove the `else`).

File: xarray/backends/api.py
Comment (mathause) [inline xarray/backends/api.py:108]: (I don't like the pattern so much but you could remove the `else`).

**Review by mathause — APPROVED** [review]
Thanks @andersy005 - I am not so sure about the inline dicts: once they begin to span more than 2 lines I find them more difficult than helpful.

**Review by keewis — COMMENTED** [review]
thanks, @andersy005.

The multiline dict comprehensions are a matter of taste (I do prefer them over item assignment in a loop where possible), but as far as I can tell the only difference is the speed of the loop evaluation in CPython (dict comprehensions iterate in C)

File: xarray/core/utils.py
Comment (mathause) [reply xarray/core/utils.py:283]: thanks - I added the point back in

File: xarray/backends/h5netcdf_.py
Comment (shoyer) [reply xarray/backends/h5netcdf_.py:104]: I'm not entirely sure, but I _think_ this is intentional. h5netcdf has a compatibility layer that defines a h5netcdf.File subclass with different semantics. Xarray only works with the base class.

(Arguably this is poor design in h5netcdf, but that's a different matter...)

File: xarray/backends/netCDF4_.py
Comment (shoyer) [inline xarray/backends/netCDF4_.py:166]: I think this was clearer before -- it's an early exit for the trivial case, rather than putting all these code in between.

File: xarray/backends/netCDF4_.py
Comment (shoyer) [reply xarray/backends/netCDF4_.py:311]: I'm less sure here...

File: xarray/coding/strings.py
Comment (shoyer) [reply xarray/coding/strings.py:133]: +1 I liked the original a little better, because it makes it more obvious that we only import dask in the block where the argument is a dask array already.

File: xarray/core/accessor_str.py
Comment (shoyer) [inline xarray/core/accessor_str.py:279]: I also don't think this is an improvement. Early `return` and one less level of indentation is help when there's a large block of code that doesn't need to be indented, but here it's only three short lines. It is easier to follow the control flow (seeing the mutually exclusive options) when the alternative is in an `else` block, so these considerations need to be balanced.

File: xarray/backends/api.py
Comment (shoyer) [inline xarray/backends/api.py:310]:
```suggestion
            f"chunks must be an int, dict, 'auto', or None. Instead found {chunks}."
```

File: xarray/backends/api.py
Comment (shoyer) [reply xarray/backends/api.py:108]: I like `else` here, it makes it easier to follow the logic :)

File: xarray/core/accessor_str.py
Comment (andersy005) [reply xarray/core/accessor_str.py]: 👍🏽

File: xarray/backends/netCDF4_.py
Comment (andersy005) [inline xarray/backends/netCDF4_.py:469]: @shoyer, unless I am missing something, this issue has been resolved upstream (https://github.com/Unidata/netcdf4-python/issues/730), right?

File: xarray/backends/h5netcdf_.py
Comment (andersy005) [reply xarray/backends/h5netcdf_.py:104]: I'm going to keep the existing `type(...)` check for the time being

File: xarray/backends/netCDF4_.py
Comment (andersy005) [reply xarray/backends/netCDF4_.py:312]: > same here: use isinstance? 

Will give it a try and see what happens 🤞🏽

**Comment by alexamici** [comment]
I feel this PR is too big and diverse to accept it as a whole.

For example I like most changes from older style formatting to f-strings, but I disagree strongly the moving the `return` statements into functions improves readability.

File: xarray/backends/api.py
Comment (alexamici) [inline xarray/backends/api.py:506]: This is an example where I think the intermediate local variable created by a multi-line statement improves readability as having `return` on the last line of the function.

Comment (max-sixty) [reply xarray/backends/api.py:506]: > I feel this PR is too big and diverse to accept it as a whole.
> 
> For example I like most changes from older style formatting to f-strings, but I disagree strongly the moving the `return` statements into functions improves readability.

While I don't have the same sense on the specifics, if people feel strongly then it may be reasonable to revert or pause on some of these.

What do you suggest re the broader PR though? While focused PRs are easier to review and reach agreement — now this is here, it does seem on net beneficial, and we should take advantage of the benefit.

Are there a few items you feel strongly about that we could revert and then merge the rest?

**Comment by max-sixty** [comment]
> I feel this PR is too big and diverse to accept it as a whole.
> 
> For example I like most changes from older style formatting to f-strings, but I disagree strongly the moving the `return` statements into functions improves readability.

While I don't have the same sense on the specifics, if people feel strongly then it may be reasonable to revert or pause on some of these.

What do you suggest re the broader PR though? While focused PRs are easier to review and reach agreement — now this is here, it does seem on net beneficial, and we should take advantage of the benefit.

Are there a few items you feel strongly about that we could revert and then merge the rest?

**Comment by alexamici** [comment]
@max-sixty & @andersy005 mine was more of a general comment. It is OK to merge it for me, but please after the release!

**Comment by andersy005** [comment]
> I feel this PR is too big and diverse to accept it as a whole.
> 
> For example I like most changes from older style formatting to f-strings, but I disagree strongly the moving the `return` statements into functions improves readability.


If folks are in favor of the original style (inline variables that are immediately returned) throughout the codebase, I'm happy to revert back to the original style :)

File: xarray/backends/netCDF4_.py
Comment (shoyer) [reply xarray/backends/netCDF4_.py:469]: Yes, it would be good to adjust the corresponding test and remove this

File: xarray/backends/h5netcdf_.py
Comment (shoyer) [inline xarray/backends/h5netcdf_.py:393]: I don't _really_ care, but I do agree with @alexamici that the separate `return` at the end is a little clearer.

File: xarray/backends/netCDF4_.py
Comment (shoyer) [inline xarray/backends/netCDF4_.py:358]: This isn't really any better from my perspective. I guess it's one less token but mostly it's just the previous stuff re-arranged.

File: xarray/backends/plugins.py
Comment (shoyer) [inline xarray/backends/plugins.py:76]: are you sure this is even doing the same check? I'm guessing the convoluted `get(be_name, None) is not None` bit was for a reason...

File: xarray/backends/zarr.py
Comment (shoyer) [inline xarray/backends/zarr.py:687]: please revert

File: xarray/coding/frequencies.py
Comment (shoyer) [inline xarray/coding/frequencies.py:240]: please revert

File: xarray/coding/strings.py
Comment (shoyer) [inline xarray/coding/strings.py:185]: please revert

File: xarray/coding/variables.py
Comment (shoyer) [inline xarray/coding/variables.py:99]: please revert

File: xarray/core/duck_array_ops.py
Comment (shoyer) [inline xarray/core/duck_array_ops.py:460]: please revert

File: xarray/core/groupby.py
Comment (shoyer) [inline xarray/core/groupby.py:645]: please revert

File: xarray/core/merge.py
Comment (shoyer) [inline xarray/core/merge.py:138]: please revert

File: xarray/core/rolling.py
Comment (shoyer) [inline xarray/core/rolling.py:436]: please revert

**Review by shoyer — CHANGES_REQUESTED** [review]

**Comment by shoyer** [comment]
@andersy005 thanks for your hard work here. On the whole, I think this will be a net benefit to the codebase, but in my review I pointed out a number of locations where I did not find the style to be improved (in my opinion). There is also one place where I think your logic may have an error.

It would be nice to come up with a policy around "style clean-up" PRs for the future. In general, I think these are _not a good idea_, for several reasons:

1. Even code that has been carefully checked can introduce bugs. Style clean-up PRs thus still need to be reviewed carefully.
2. Fewer lines or tokens of code is not always better. Good examples from this PR include reduced nesting with early `return` statements, comprehensions and conditional expressions. All of these would be unobjectionable to use in new code (and are often a good idea), but code that doesn't use them even when it could is fine, too.
3. Some ugly code works fine, and would not benefit from clean-up. For example, `xarray.util.print_versions` (which was forked from pandas) is quite ugly code. But it's peripheral to the project, and has barely been touched aside from automated reformatting.

Instead, I would suggest a policy of a _selective cleanup only_. Feel free to clean-up style when it arises in the process of normal coding, e.g., when you are modifying nearly or related code for another reason. This reduces the risk of introducing inadvertent errors, and the fact that you were working on nearby code means that this part of the codebase is worthy of improvement.

Code-base wide clean-up for particular issues may be warranted occasionally, but I would suggest always discussing it with other core developers first.

File: xarray/backends/netCDF4_.py
Comment (alexamici) [reply xarray/backends/netCDF4_.py:358]: Also, since this is now a single line, it is harder to tell if both branches are tested.

File: xarray/backends/plugins.py
Comment (keewis) [reply xarray/backends/plugins.py:76]: it seems my mistake was to assume that entries to `backend_entrypoints` are never equal to `None`. I'm not sure why that would happen, though (maybe `Entrypoint.load` returns `None` if it fails? But couldn't we just not assign then?). If that's not a mistake I would probably remove all `None` valued entries before this because it's much easier to understand if we don't do multiple things at the same time.

File: xarray/backends/netCDF4_.py
Comment (andersy005) [reply xarray/backends/netCDF4_.py:469]: Okay.. Will address this in a separate PR.

**Comment by andersy005** [comment]
@shoyer, Thank you for the thorough feedback. 

> Even code that has been carefully checked can introduce bugs. Style clean-up PRs thus still need to be reviewed carefully.

> Fewer lines or tokens of code is not always better. Good examples from this PR include reduced nesting with early return statements, comprehensions and conditional expressions. All of these would be unobjectionable to use in new code (and are often a good idea), but code that doesn't use them even when it could is fine, too.

👍🏽. I reverted back to the original style in most places. There're a few comments that I haven't addressed yet. I will look into those sometime this weekend. 


> Code-base wide clean-up for particular issues may be warranted occasionally, but I would suggest always discussing it with other core developers first.

I agree. I didn't expect the PR to get this large. My original intent when I created the PR was to clean the backends code, and I ended up invading other areas  :(. In the future, I'll definitely make sure to start a discussion before working on changes that may end up affecting a large chunk of the code base.

**Comment by max-sixty** [comment]
This is going to collect merge conflicts — shall we merge?

**Comment by andersy005** [comment]
@max-sixty, I'm going to address a few issues pointed out during the review, then we can merge afterwards.

**Comment by dcherian** [comment]
Shall we also add parts of Stephan's comment to the "Contributing Guide" in the documentation?

File: xarray/backends/plugins.py
Comment (andersy005) [reply xarray/backends/plugins.py]: I reverted back to the original version so as to avoid accidentally breaking things :)

**Comment by andersy005** [comment]
> Shall we also add parts of Stephan's comment to the "Contributing Guide" in the documentation?

👍🏽 for addressing this in a separate PR

File: xarray/core/accessor_dt.py
Comment (andersy005) [reply xarray/core/accessor_dt.py]: I ran into a couple errors when I tried this change :(. I am going to leave this as is for the time being.

**Comment by andersy005** [comment]
I believe I have addressed most, if not all changes requested during the review. If there are still disputable changes that need addressing, please let me know

File: xarray/coding/variables.py
Comment (shoyer) [inline xarray/coding/variables.py:89]: these new lines should probably be kept :)

**Review by shoyer — APPROVED** [review]

**Comment by dcherian** [comment]
Looks like this is done. Thanks @andersy005

**Comment by max-sixty** [comment]
Thanks @andersy005 !

**Pull request merged**
